client.go

```diff
@@ -1,48 +1,273 @@
 		return nil, err
 	}
 
-	// Initialise amazon client with internal client.
-	amzCli := &AmzClientImpl{
-		goamzs3.New(auth, aws.Regions[region]),
+	// Create AWS-SDK-S3 client with the session
+	sdkClient := s3.New(sess)
+
+	// If we require crypto client (HasUserDefinedPSK), create it.
+	var cryptoClient S3CryptoClient
+	if HasUserDefinedPSK {
+		cryptoConfig := &s3crypto.Config{HasUserDefinedPSK: true}
+		s3cryptoClient := s3crypto.New(sess, cryptoConfig)
+
+		cryptoClient = s3cryptoClient
 	}
 
-	// Create S3 with the created amzClient.
-	return NewWithClient(amzCli, bucketName), nil
+	cli := Instantiate(sdkClient, cryptoClient, bucketName, region)
+	return cli, nil
 }
 
-// NewWithClient returns a new S3 structure for the provided AmzClient instance.
-func NewWithClient(client AmzClient, bucketName string) *S3 {
+// Instantiate creates a new instance of S3 struct with the provided clients, bucket and region
+func Instantiate(sdkClient S3SDKClient, cryptoClient S3CryptoClient, bucketName, region string) *S3 {
 	return &S3{
-		cli:        client,
-		BucketName: bucketName,
+		sdkClient:    sdkClient,
+		cryptoClient: cryptoClient,
+		bucketName:   bucketName,
+		region:       region,
 	}
 }
 
-// Get returns an io.ReadCloser instance for the given fully qualified S3 URL.
-// Note that this function will mutate the bucket name of the S3 object if it's different than the existing one.
-func (s3 *S3) Get(rawURL string) (io.ReadCloser, error) {
+// BucketName is a getter for the bucket name used by this S3 client
+func (cli *S3) BucketName() string {
+	return cli.bucketName
+}
+
+// Upload handles the uploading a file to AWS S3, into the bucket configured for this client
+func (cli *S3) Upload(ctx context.Context, req *UploadRequest, payload []byte) error {
+	return cli.UploadWithPsk(ctx, req, payload, nil)
+}
+
+// UploadWithPsk handles the uploading a file to AWS S3, into the bucket configured for this client, using a user-defined psk
+func (cli *S3) UploadWithPsk(ctx context.Context, req *UploadRequest, payload []byte, psk []byte) error {
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		log.Event(ctx, "error fetching multipart list", log.Error(err))
+		return err
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+		}
+	}
+
+	if len(uploadID) == 0 {
+
+		createMultiInput := &s3.CreateMultipartUploadInput{
+			Bucket:      &cli.bucketName,
+			Key:         &req.UploadKey,
+			ContentType: &req.Type,
+		}
+
+		createMultiOutput, err := cli.sdkClient.CreateMultipartUpload(createMultiInput)
+		if err != nil {
+			log.Event(ctx, "error creating multipart upload", log.Error(err))
+			return err
+		}
+
+		uploadID = *createMultiOutput.UploadId
+	}
+
+	rs := bytes.NewReader(payload)
+
+	n := int64(req.ChunkNumber)
```

Comment:
Is it possible this should fail the cast to int64?

Reply:
I copied this from Florence repo, I think it makes more sense to enforce int64 type in the UploadRequest struct (provided by the caller) instead of casting it here. I will do this change.